compiler/rustc_mir_transform/src/fold.rs

```diff
@@ -0,0 +1,357 @@
+
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        while did_optimization(body, tcx) {}
+    }
+}
+
+// Fold:
+// _2 = &_1;
+// _3 = *_2;
+// Into:
+// _3 = _1;
+fn fold_ref_deref<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    _temp_place: &Place<'tcx>,
+    temp_rvalue: &Rvalue<'tcx>,
+    final_place: &Place<'tcx>,
+    final_rvalue: &Rvalue<'tcx>,
+) -> Option<StatementKind<'tcx>> {
+    let Rvalue::Ref(_, _, first_place) = temp_rvalue else { return None; };
+    let Rvalue::Use(final_operand) = final_rvalue else { return None; };
+
+    let Some(second_place) = final_operand.place() else { return None; };
+    if second_place.projection.get(0) != Some(&ProjectionElem::Deref) {
+        return None;
+    }
+
+    // Combine all the projections together
+    let new_projections: Vec<_> =
+        first_place.projection.iter().chain(second_place.projection.iter().skip(1)).collect();
+
+    // Indexing projections require that we have the Index live
+    // Ensuring that is complicated. Bail.
+    if new_projections.iter().any(|p| matches!(p, ProjectionElem::Index(_))) {
+        return None;
+    }
+
+    let new_place = Place {
+        local: first_place.local,
+        projection: tcx.intern_place_elems(&new_projections[..]),
+    };
+
+    if new_place == *final_place {
+        Some(StatementKind::Nop)
+    } else {
+        Some(StatementKind::Assign(Box::new((*final_place, Rvalue::Use(Operand::Copy(new_place))))))
+    }
+}
+
+// Fold:
+// _2 = _1.a;
+// _3 = _2.b;
+// Into:
+// _3 = _1.a.b;
+fn fold_place_projections<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    temp_place: &Place<'tcx>,
+    temp_rvalue: &Rvalue<'tcx>,
+    final_place: &Place<'tcx>,
+    final_rvalue: &Rvalue<'tcx>,
+) -> Option<StatementKind<'tcx>> {
+    let Rvalue::Use(temp_operand) = temp_rvalue else { return None; };
+    let Rvalue::Use(final_operand) = final_rvalue else { return None; };
+
+    // If we are assigning into a place expression, that would be something like
+    // _3.b = _2.b;
+    // Which is complicated. Just don't optimize that at all for now.
+    if !temp_place.projection.is_empty() {
+        return None;
+    }
+
+    // If either of these places are constants, ouch
+    let Some(first_place) = temp_operand.place() else { return None; };
+    let Some(second_place) = final_operand.place() else { return None; };
+
+    // Derefs must come first if at all
+    // If merging these assignments would break that rule, bummer. Bail.
+    if first_place.projection.len() > 0
+        && second_place.projection.get(0) == Some(&ProjectionElem::Deref)
+    {
+        return None;
+    }
+
+    // Combine all the projections together
+    let new_projections: Vec<_> =
+        first_place.projection.iter().chain(second_place.projection.iter()).collect();
+
+    // Indexing projections require that we have the Index live
+    // Ensuring that is complicated. Bail.
+    if new_projections.iter().any(|p| matches!(p, ProjectionElem::Index(_))) {
```

Comment:
It's only complicated if `first_place` has `Index`. The projections in `second_place` don't move, so they should still be correct.

Reply:
Fair point. I'd still rather solve this by improving my liveness ranges hack :p